File: ring.js
var cluster = require('cluster')
  , http = require('http')

module.exports = function (n, m) {
  if (cluster.isMaster) {
    console.time('in')
    var worker, ni = n, mi = m
    while (ni--) worker = new Worker(worker)
    while (mi--) worker.hello()

    cluster.on('exit', function (worker, code, signal) {
      if (!Object.keys(cluster.workers).length) {
        console.log('%d messages sent through ring', n * m)
        console.timeEnd('in')
      }
    })
  } else {
    var mCount = 0
    process.on('message', function (msg) {
      if (++mCount >= m) {
        cluster.worker.kill()
      }
    })
  }
}

function Worker(next) {
  var me = cluster.fork()
  me.hello = function () {
    me.send('hello')
    if (next) {
      next.hello()
    }
  }
  return me
}

File: ring.erl
-module(ring).
-export([start/2]).

start(N, M) ->
  statistics(runtime),
  start(N, M, N, spawn_link(fun() -> loop(N, M, []) end)).

start(_, M, 0, Next) ->
  send(Next, hey, M);
start(N, M, NC, Next) ->
  start(N, M, NC-1, spawn_link(fun() -> loop(N, M, Next) end)).

send(Pid, _, 0) ->
  Pid ! done,
  Pid;
send(Pid, Message, M) ->
  Pid ! Message,
  send(Pid, Message, M-1).

loop(N, M, Next) ->
  receive
    done when is_pid(Next) ->
      Next ! done;
    done ->
      Z = N * M,
      {_, Time} = statistics(runtime),
      io:format("~p messages sent in ~p ms~n", [Z, Time]),
      exit(self(), ok);
    hey when is_pid(Next) ->
      Next ! hey,
      loop(N, M, Next);
    _ ->
      loop(N, M, Next)
  end.
